### PR TITLE
Controlnet

### PR DESCRIPTION
**Pull request by RomainDeleat-besson**
Hello,

Thank you for taking the time to review my post !

### CODE REVIEW

File: collections/_posts/2025-03-26-ControlNet.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+Diffusion models can create visually stunning images by typing in a text prompt, but they are limited in the control they provide over the spatial composition of the image.
+In particular, it is difficult to accurately express complex layouts and poses using text prompts alone.
+
+ControlNet is a neural network architecture that can add spatial control (e.g. edge maps, human pose skeletons, segmentation maps, depth, normals, etc.) to large pre-trained text-to-image diffusion models.
+
+
+
+# Methods
+
+ControlNet is a neural network architecture designed to maintain the quality and robustness of a large pretrained model, specifically a text-to-image Stable Diffusion model [1].
```
Comment (Tmodrzyk) [inline collections/_posts/2025-03-26-ControlNet.md:33]: Reiterate that the aim is to preserve image quality **by adding more control over composition**, as is done in the intro. Otherwise the sentence is ambiguous.

File: collections/_posts/2025-03-26-ControlNet.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+* $$ \mathcal{Z}(.;.) $$ is a $$ 1 \times 1 $$ convolution layer with both weights and biases initialized to zeros.
+* $$ \Theta_{z1} $$ and $$ \Theta_{z2} $$ are the parameters of the first and second *zero convolution* layer.
+
+
+<div style="text-align:center"><img src="/collections/images/ControlNet/Fig_1bis.jpg" width=1500></div>
+
+
+
+The objective function is:
```
Comment (Tmodrzyk) [inline collections/_posts/2025-03-26-ControlNet.md:55]: Perhaps add an explanation (even a simple one) of why we have this loss and what the different terms do?

File: collections/_posts/2025-03-26-ControlNet.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+
+# Results
+
+## CFG Resolution Weighting and Composition
+
+<div style="text-align:center"><img src="/collections/images/ControlNet/Fig_2.jpg" width=1500></div>
+
+Classifier-free guidance was used in Stable Diffusion model to generate high quality images.
+It is formulated as:
```
Comment (Tmodrzyk) [inline collections/_posts/2025-03-26-ControlNet.md:75]: Here, too, you might want to explain the purpose of the different terms in the loss and why they're there.

File: collections/_posts/2025-03-26-ControlNet.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+$$ \epsilon_{\text{prd}} = \epsilon_{\text{uc}} + \beta_{\text{cfg}} (\epsilon_c - \epsilon_{\text{uc}}) $$ 
+
+* $$ \epsilon_{\text{prd}} $$ is the model’s final output.
+* $$ \epsilon_{\text{uc}} $$ is the unconditional output.
+* $$ \epsilon_c $$ is the conditional output.
+* $$ \beta_{\text{cfg}} $$ is user-specified weight.
+
+
+They had to change the original CFG (as shown in figure 5) by adding the conditioning image to $$ \epsilon_c $$ and multiply a weight $$ w_i $$ (based on the resolution of each block) to each connection between Stable Diffusion and ControlNet.
```
Comment (Tmodrzyk) [inline collections/_posts/2025-03-26-ControlNet.md:85]: It's not clear to me why they need to do this, even if you explain afterwards what impact it had.

File: collections/_posts/2025-03-26-ControlNet.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+
+Quantitative results:
+
+<div style="text-align:center"><img src="/collections/images/ControlNet/Tab_1.jpg" width=1500></div>
+
+
+## Dataset size and Input interpretation
+
+<div style="text-align:center"><img src="/collections/images/ControlNet/Fig_4.jpg" width=1500></div>
```
Comment (Tmodrzyk) [inline collections/_posts/2025-03-26-ControlNet.md:106]: Maybe add a little explanation paragraph to go along the figure

File: collections/_posts/2025-03-26-ControlNet.md (hunk 1 above)
Comment (RomainDeleat-besson) [reply collections/_posts/2025-03-26-ControlNet.md]: ok, I changed the sentence by:

ControlNet is a neural network architecture designed to add more spatial control while maintaining the quality and robustness of a large pretrained model (specifically in the case of the text-to-image model from Stable Diffusion model [1]).

File: collections/_posts/2025-03-26-ControlNet.md (hunk 2 above)
Comment (RomainDeleat-besson) [reply collections/_posts/2025-03-26-ControlNet.md]: Yes, here is the new sentence:

The objective function aims to minimize the original noise with the estimated noise by computing:

File: collections/_posts/2025-03-26-ControlNet.md (hunk 3 above)
Comment (RomainDeleat-besson) [reply collections/_posts/2025-03-26-ControlNet.md]: I changed the sentence by:

Classifier-free guidance was used in Stable Diffusion model to generate high quality images.
To generate these high quality images, they predicted a noise from a conditional and unconditional noise and add them using the formula below:

File: collections/_posts/2025-03-26-ControlNet.md (hunk 4 above)
Comment (RomainDeleat-besson) [reply collections/_posts/2025-03-26-ControlNet.md]: I tried to be a little more clear with the sentence after this paragraph:

By doing so, they reduced the guidance of the conditioning that is too strong and called their approch: CFG Resolution Weighting.

File: collections/_posts/2025-03-26-ControlNet.md (hunk 5 above)
Comment (RomainDeleat-besson) [reply collections/_posts/2025-03-26-ControlNet.md:109]: Here is the added paragraph:

The impact of the dataset size as well as an example of the network to interpret the content of an input conditioning with low context:

**Comment by RomainDeleat-besson** [comment]
Thank you for your comments Thibaut!

I have modified the post accordingly.

Please, let me know if you have further comments.

**Review by nathanpainchaud — APPROVED** [review]
